File: csharp/Snet/Rewriter.cs
using System;
using System.IO;

namespace Snet
{
	internal class Rewriter
	{
		private byte[] _Data;
		private int _head;

		private int _len;

		public Rewriter (int size)
		{
			_Data = new byte[size];
		}

		public void Push(byte[] b, int offset, int size) {
			if (size >= _Data.Length) {
				int drop = size - _Data.Length;

				Buffer.BlockCopy (b, offset + drop, _Data, 0, size - drop);
				_head = 0;
				if (_len != _Data.Length){
					_len = _Data.Length;
				}
			} else {
				int space = _Data.Length - _head;
				if (space >= size){
					Buffer.BlockCopy(b, offset, _Data, _head, size);
					if (space == size){
						_head = 0;
					} else{
						_head += size;
					}

					if (_len != _Data.Length){
						_len = Math.Min(_len + size, _Data.Length);
					}
				} else{
					Buffer.BlockCopy(b, offset, _Data, _head, space);
					Buffer.BlockCopy(b, offset + space, _Data, 0, size - space);
					_head = size - space;

					if (_len != _Data.Length){
						_len = _Data.Length;
					}
				}
			}
		}

		public bool Rewrite(Stream stream, ulong writeCount, ulong readCount) {
			int n = (int)writeCount - (int)readCount;

			if (n == 0) {
				return true;
			} else if (n < 0 || n > _len) {
				return false;
			}

			int offset = _head - n;
			if (offset >= 0){
				stream.Write(_Data, offset, n);
			} else{
				offset += _Data.Length;
				stream.Write(_Data, offset, _Data.Length - offset);
				stream.Write(_Data, 0, _head);
			}
			return true;
		}
	}
}

File: go/rewriter.go
package snet

import (
	"io"
)

type rewriter struct {
	data   []byte
	head   int
	length int
}

func (r *rewriter) Push(b []byte) {
	if len(b) >= len(r.data) {
		drop := len(b) - len(r.data)
		copy(r.data, b[drop:])
		r.head, r.length = 0, len(r.data)
		return
	}

	size := copy(r.data[r.head:], b)

	remain := len(b) - size

	if remain == 0 {
		r.head += size
		if r.head == len(r.data) {
			r.head = 0
		}

		if r.length != len(r.data) {
			r.length += len(r.data)
		}
	} else {
		r.head = copy(r.data, b[size:])
		if r.length != len(r.data) {
			r.length = len(r.data)
		}
	}
}

func (r *rewriter) Rewrite(w io.Writer, writeCount, readCount uint64) bool {
	n := int(writeCount - readCount)

	switch {
	case n == 0:
		return true
	case n < 0 || n > r.length:
		return false
	case n <= r.head:
		_, err := w.Write(r.data[r.head-n : r.head])
		return err == nil
	}

	offset := r.head - n + len(r.data)
	if _, err := w.Write(r.data[offset:]); err != nil {
		return false
	}

	_, err := w.Write(r.data[:r.head])
	return err == nil
}
